app/models/concerns/exportable.rb

```diff
@@ -0,0 +1,71 @@
+# This concern makes it easy to export an ActiveRecord object with specified
+# attributes and associations in a particular format.  If a specified
+# assocation also includes this concern, then it will likewise be exported.
+#
+# Custom attributes can be specified with a custom export lambda in an options
+# hash.
+#
+# Example:
+#   attr_exportable :name, :address, :company => ->(user) { user.company.name }
+#
+module Exportable
+  extend ActiveSupport::Concern
+
+  def to_export(include_blank = true)
+    return unless self.class.exportable_attributes.present?
+
+    self.class.exportable_attributes.keys.inject({}) do |hash, attribute|
+      value = export_attr(attribute, self.class.exportable_attributes[attribute], include_blank)
+
+      # Rails considers false blank, but if a boolean value is explicitly set false, we want to ensure we export it.
+      if include_blank || value.present? || value == false
+        hash.update(attribute => value)
+      else
+        hash
+      end
+    end.stringify_keys
+  end
+
+  # Export a particular attribute or association.
+  #   - If our exportable_attributes value is callable, we call it with self as an argument
+  #   - If our object is iterable, then we export each item
+  #   - If the attribute or association also includes this concern, call to_export on it
+  def export_attr(attribute, exporter, include_blank)
+    value = if exporter.respond_to?(:call)
+              exporter.call(self)
+            elsif self.respond_to?(exporter)
+              self.send(exporter)
+            end
+
+    value = value.respond_to?(:map) ? export_iterable(value, include_blank) : value
+    value.respond_to?(:to_export) ? value.to_export(include_blank) : value
+  end
+
+  # Exports each item in an iterable.  If it's a hash, then export each value.
+  def export_iterable(items, include_blank)
+    if items.is_a?(Hash)
+      items.each { |key, value| items[key] = value.respond_to?(:to_export) ? value.to_export(include_blank) : value }
+      items.to_hash.stringify_keys
+    else
+      items.map { |item| item.respond_to?(:to_export) ? item.to_export(include_blank) : item }
+    end
+  end
+
+  module ClassMethods
+    attr_reader :exportable_attributes
```

Comment:
This definition does not work with STI correctly since instance variable is used (on class level). When this is stored in `Subnet` object, `Subnet::IPv4`'s `@exportable_attributes` is nil. I'd recommend defining the reader so that it loads the values from parent class (if it responds to it) and add it to its own @exportable_attributes.

Or, both `Subnet::IPv4` and `Subnet::IPv6` would have to call attr_exportable with parent keys fetched recursively. With correctly used inheritance I think the first solution is better.

OT: @@exportable_attributes would be shared but if child class would add something, parent would get it as well so we shouldn't use that